File: solutions/clojure/matching-brackets/1/src/matching_brackets.clj
(ns matching-brackets)

(defn valid? [s]
  (loop [stack ()
         [c & rs] s]
    (cond (nil? c) (empty? stack)
          (#{\( \[ \{} c) (recur (conj stack c) rs)
          (#{\) \] \}} c) (if (not= c ({\( \) \[ \] \{ \}} (first stack)))
                            false
                            (recur (rest stack) rs))
          :else (recur stack rs))))

File: solutions/go/matching-brackets/1/matching_brackets.go
package brackets

func Bracket(input string) bool {
	stack := []rune{}
    for _, ch := range input {
        switch ch {
            case '(', '[', '{':
            	stack = append(stack, ch)
            case ')':
            	if l := len(stack); l == 0 || stack[l-1] != '(' {
                    return false
                }
            stack = stack[:len(stack)-1]
            case ']':
            	if l := len(stack); l == 0 || stack[l-1] != '[' {
                    return false
                }
            stack = stack[:len(stack)-1] 
            case '}':
            	if l := len(stack); l == 0 || stack[l-1] != '{' {
                    return false
                }
            stack = stack[:len(stack)-1]
        }
    }
    return len(stack) == 0
}
